README.md
# tailwind-typewriter

A plugin for [Tailwind CSS](https://tailwindcss.com/) that generates typewriter style text
animations in pure CSS _(no JavaScript required)_ ✨

### Support

This plugin has been **tested with Tailwind CSS v3**. Previous releases may also work but are not
officially supported.

## Demo

![Demo](./.github/demo.gif)

## Installation

Install the plugin:

```sh
# Using npm
npm install tailwind-typewriter

# Using Yarn
yarn add tailwind-typewriter
```

Then add the plugin to your `tailwind.config.js` file:

```js
// tailwind.config.js
module.exports = {
    theme: {
        // ...
    },
    plugins: [
        require('tailwind-typewriter')
        // ...
    ]
};
```

## Usage

The quickest, dirtiest way to get started is to insert a `<span>` with the `type-example` utility
class somewhere in your markup. This will let you verify that everything is working with a built-in
example:

```html
<span class="type-example"></span>
```

### Styling

You can style the animation text with the usual Tailwind utilities:

```html
<span class="type-example text-lg text-rose-500"></span>
```

<small>🗒️ **Note:** The blinking caret can only be styled using the configuration options
below.</small>

### Customizing

The above example is essentially useless on its own. In a real world application, you'll probably
want to provide your own text etc 😎. If, for example, you want to create an animation that cycles
between names of fruit (with a 3 second delay) - you can do the following:

```js
// tailwind.config.js
module.exports = {
    theme: {
        // ...
    },
    plugins: [
        require('tailwind-typewriter')({
            wordsets: {
                fruit: {
                    words: ['apple', 'banana', 'orange', 'pear', 'strawberry'],
                    delay: 3
                }
            }
        })
    ]
};
```

```html
<span class="type-fruit"></span>
```

<small>🗒️ **Note:** The animation is generated in the `::after` pseudo-element so it will be
appended to any text inserted inside the `<span>`. This may or may not be desirable depending on
your project.</small>

#### Multiple Animations

You can create multiple animations in a single project. Each animation is called a 'wordset'. Each
wordset has a name, a set of words/phrases, and its own configuration options
([see the full API below](#configuration)).

```js
// tailwind.config.js
module.exports = {
    theme: {
        // ...
    },
    plugins: [
        require('tailwind-typewriter')({
            wordsets: {
                fruit: {
                    words: ['apple', 'banana', 'orange', 'pear', 'strawberry'],
                    delay: 3
                },
                vegetables: {
                    words: ['carrot', 'celery', 'corn', 'potato'],
                    pauseBetween: 5
                }
            }
        })
    ]
};
```

A unique class is generated for each set as `type-{wordset}`. For example, to reference the
_Vegetables_ animation instead of _Fruit_, you would use the following:

```html
<span class="type-vegetables"></span>
```

#### Singular Animations

This plugin was originally designed to create animations that cycled through a set of words.
Sometimes however, you may need to create a simple effect for a single word or phrase. You probably
also want this animation to output in a forwards direction, stopping after the last character.

To do this, choose your phrase, then set `repeat` and `eraseSpeed` to `0`:

```js
// tailwind.config.js
module.exports = {
    theme: {
        // ...
    },
    plugins: [
        require('tailwind-typewriter')({
            wordsets: {
                fruit: {
                    words: ['Your clever phrase'],
                    repeat: 0,
                    eraseSpeed: 0
                }
            }
        })
    ]
};
```

#### A word about performance

Generally speaking, CSS animations are very performant (certainly compared with JavaScript).
However, it is worth noting that every letter in a wordset generates a CSS keyframe. This means that
if you have a wordset with a large amount of text, this will have some impact on file size and
possibly renderer performance. This wont be a problem in most use cases, but in large projects,
discretion is strongly advised.

## Configuration

### `wordsets`

Currently, this is the only available top-level property. It serves as a wrapper for each wordset
grouping. A wordset in essence, is just an object literal with a unique name for the key that
contains the necessary configuration options for each animation:

```js
wordsets: {
    fruit: {
        words: ['apple', 'banana', 'orange'],
        // ..
    },
    vegetables: {
        words: ['carrot', 'celery', 'corn'],
        // ...
    }
}
```

### Options

The following options are also available to each individual wordset.

| Option         | Description                                                                | Accepts     | Default            |
| :------------- | :------------------------------------------------------------------------- | ----------- | ------------------ |
| `words`        | A list of words or phrases for each step of the animation.                 | `Array`     | _Built-in Example_ |
| `delay`        | Delays the start of the animation by _**n** seconds_.                      | `Number`    | `1`                |
| `repeat`       | Repeat the animation _**n** times_. `-1` for an infinite loop.             | `Number`    | `-1`               |
| `writeSpeed`   | Speed per letter during the write phase _(**n** seconds)_.                 | `Number`    | `0.3`              |
| `eraseSpeed`   | Speed per letter during the erase phase _(**n** seconds)_. `0` to disable. | `Number`    | `0.1`              |
| `pauseBetween` | Pause for _**n** seconds_ between each word.                               | `Number`    | `4`                |
| `caretColor`   | Set the color of the caret.                                                | `CSS Color` | `'currentColor'`   |
| `caretWidth`   | Specify an alternate width for the caret.                                  | `CSS Unit`  | `'1px'`            |
| `caretSpacing` | Space between the last character and the caret.                            | `CSS Unit`  | `'0.1em'`          |
| `blinkSpeed`   | The frequency at which the caret blinks _(**n** seconds)_.                 | `Number`    | `0.8`              |


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.1](https://github.com/mattpfeffer/tailwind-typewriter/compare/v1.0.0...v1.0.1) (2022-08-03)


### Bug Fixes

* **package:** include compiled entry file ([82aa74c](https://github.com/mattpfeffer/tailwind-typewriter/commit/82aa74ce91c2716fd0e714cee804277ecf1c3fd9))

## 1.0.0 (2022-08-03)


test/index.test.ts
const postcss = require('postcss');
const typewriterPlugin = require('../dist/index');

/*
 * Configuration
 */

// Mock Options
const options = {
    name: 'vegetables',
    words: ['corn', 'carrot', 'potato'],
    delay: Math.PI,
    repeat: 0,
    writeSpeed: 2,
    eraseSpeed: 1,
    pauseBetween: 3,
    caretColor: 'white',
    caretWidth: '2px'
};

/*
 * Helper Functions
 */

describe('Helper Functions', () => {
    test('round() -> rounds a number to two decimal places', () => {
        const round = typewriterPlugin.__get__('round');

        // Test number input (use c.delay for generic input)
        expect(round(options.delay)).toBe(3.14);

        // Test string input
        expect(round(String(options.delay))).toBe(3.14);
    });

    test('randomize() -> adds a small random variance to a number', () => {
        const randomize = typewriterPlugin.__get__('randomize');

        const number = options.delay;
        const string = String(number);
        const variance = 0.5;

        let result;

        // Test number input
        result = randomize(number, variance);
        expect(result).toBeGreaterThanOrEqual(number - variance);
        expect(result).toBeLessThanOrEqual(number + variance);
        expect(result).not.toBe(number);

        // Test string input
        result = randomize(string, variance);
        expect(result).toBeGreaterThanOrEqual(number - variance);
        expect(result).toBeLessThanOrEqual(number + variance);
        expect(result).not.toBe(number);
    });
});

/*
 * Core Functions
 */

describe('Core Functions', () => {
    test('generateKeyframes() -> generates keyframes for a given wordset', () => {
        const generateKeyframes = typewriterPlugin.__get__('generateKeyframes');

        const expected = {
            [`@keyframes write-${options.name}`]: {
                from: { content: expect.stringMatching(/""/) },
                to: { content: expect.stringMatching(/""|"\w+"/) }
            }
        };

        const sample = expect.stringMatching(/"\d+(?:\.\d+)?%":\{"content":"\\"\w*\\""\},/);

        let result;

        // Test default usage
        result = generateKeyframes(
            options.name,
            options.words,
            options.writeSpeed,
            options.eraseSpeed,
            options.pauseBetween,
            options.repeat === -1
        );

        expect(result).toMatchObject(expected);
        expect(JSON.stringify(result).replace('\\', '')).toEqual(sample);

        // Re-test with eraseSpeed disabled
        result = generateKeyframes(
            options.name,
            options.words,
            options.writeSpeed,
            0,
            options.pauseBetween,
            options.repeat === -1
        );

        expect(result).toMatchObject(expected);
        expect(JSON.stringify(result).replace('\\', '')).toEqual(sample);
    });

    test('animationLength() -> calculates the total animation time in seconds', () => {
        const animationLength = typewriterPlugin.__get__('animationLength');

        const result = animationLength(
            options.words,
            options.writeSpeed,
            options.eraseSpeed,
            String(options.pauseBetween)
        );

        expect(result).toBe(57);
    });

    test('calculateSteps() -> calculates keyframe steps for each word phase', () => {
        const calculateSteps = typewriterPlugin.__get__('calculateSteps');

        const result = calculateSteps(options.words[0], options.writeSpeed, 50);

        expect(result.length).toBe(5);
        expect(result).toContain(12);
    });
});

/*
 * Plugin Declaration
 */

// Default config test strings
const defaultOutput = {};

defaultOutput.a = `
@keyframes blink-caret-example {
    from, to {
        border-color: transparent
    }
    50% {
        border-color: currentColor
    }
}
`;

defaultOutput.b = `
@keyframes write-example {
    from {
        content: ""
    }
    to {
        content: ""
    }
`;

defaultOutput.c = 'content: "bananas"';

/* eslint-disable no-useless-escape */
defaultOutput.d = `
.type-example::after {
    content: \"\";
    border-right: 1px solid currentColor;
    padding-right: 0.1em;
    animation: blink-caret-example 0.8s infinite, write-example 20s 1s infinite forwards;
    transition: border-color 50ms cubic-bezier(0.4, 0, 0.2, 1)
}
`;
/* eslint-enable no-useless-escape */

// User config test strings
const userOutput = {};

userOutput.a = `
@keyframes blink-caret-${options.name} {
    from, to {
        border-color: transparent
    }
    50% {
        border-color: ${options.caretColor}
    }
}
`;

userOutput.b = `
@keyframes write-${options.name} {
    from {
        content: ""
    }
    to {
        content: ""
    }
`;

userOutput.c = `content: "${options.words[2]}"`;

/* eslint-disable no-useless-escape */
userOutput.d = `
.type-${options.name}::after {
    content: \"\";
    border-right: ${options.caretWidth} solid ${options.caretColor};
    padding-right: 0.1em;
    animation: blink-caret-${options.name} 0.8s infinite, write-${options.name} 57s ${
    options.delay
}s ${options.repeat + 1} forwards;
    transition: border-color 50ms cubic-bezier(0.4, 0, 0.2, 1)
}
`;
/* eslint-enable no-useless-escape */

describe('Plugin Declaration', () => {
    test('with default options -> generates keyframes and utility classes', () => {
        const result = postcss([
            require('tailwindcss')({
                content: [
                    {
                        raw: 'type-example'
                    }
                ],
                plugins: [require('../dist')]
            })
        ]).process('@tailwind utilities').css;

        expect(result).toEqual(expect.stringContaining(defaultOutput.a.trim()));
        expect(result).toEqual(expect.stringContaining(defaultOutput.b.trim()));
        expect(result).toEqual(expect.stringContaining(defaultOutput.c.trim()));
        expect(result).toEqual(expect.stringContaining(defaultOutput.d.trim()));
    });

    test('with basic options -> generates keyframes and utility classes', () => {
        const result = postcss([
            require('tailwindcss')({
                content: [
                    {
                        raw: `type-${options.name}`
                    }
                ],
                plugins: [
                    require('../dist')({
                        wordsets: {
                            [options.name]: {
                                words: options.words
                            }
                        }
                    })
                ]
            })
        ]).process('@tailwind utilities').css;

        expect(result).toEqual(expect.stringContaining(userOutput.b.trim()));
        expect(result).toEqual(expect.stringContaining(userOutput.c.trim()));
    });

    test('with advanced options -> generates keyframes and utility classes', () => {
        const result = postcss([
            require('tailwindcss')({
                content: [
                    {
                        raw: `type-${options.name}`
                    }
                ],
                plugins: [
                    require('../dist')({
                        wordsets: {
                            [options.name]: options
                        }
                    })
                ]
            })
        ]).process('@tailwind utilities').css;

        expect(result).toEqual(expect.stringContaining(userOutput.a.trim()));
        expect(result).toEqual(expect.stringContaining(userOutput.b.trim()));
        expect(result).toEqual(expect.stringContaining(userOutput.c.trim()));
        expect(result).toEqual(expect.stringContaining(userOutput.d.trim()));
    });
});


src/index.ts
import type { Wordset } from './types';

const num = require('d3-interpolate');
const esc = require('cssesc');

const plugin = require('tailwindcss/plugin');

/*
 * Plugin Declaration
 */

// @ts-ignore
module.exports = plugin.withOptions((options) => {
    // @ts-ignore
    return function ({ addUtilities }) {
        // Set defaults
        const d = {
            words: ['apples', 'oranges', 'bananas'],
            delay: 1,
            repeat: -1,
            writeSpeed: 0.3,
            eraseSpeed: 0.1,
            pauseBetween: 4,
            caretColor: 'currentColor',
            caretWidth: '1px',
            caretSpacing: '0.1em',
            blinkSpeed: 0.8
        };

        // Get options
        let wordsets;
        if (typeof options !== 'undefined') {
            wordsets = options.wordsets;
        } else {
            wordsets = {
                example: d
            };
        }

        const utilities = [
            // Define caret animation
            Object.entries(wordsets).map(([key, wordset]) => {
                const { caretColor = d.caretColor } = wordset as Wordset;
                return {
                    [`@keyframes blink-caret-${esc(key, {
                        isIdentifier: true
                    })}`]: {
                        'from, to': { 'border-color': 'transparent' },
                        '50%': { 'border-color': `${caretColor}` }
                    }
                };
            }),

            // Generate text animations
            Object.entries(wordsets).map(([key, wordset]) => {
                const {
                    words,
                    repeat = d.repeat,
                    writeSpeed = d.writeSpeed,
                    eraseSpeed = d.eraseSpeed,
                    pauseBetween = d.pauseBetween
                } = wordset as Wordset;

                return generateKeyframes(
                    key,
                    words,
                    writeSpeed,
                    eraseSpeed,
                    pauseBetween,
                    repeat === -1
                );
            }),

            // Define utility classes
            Object.entries(wordsets).map(([key, wordset]) => {
                const {
                    words,
                    delay = d.delay,
                    repeat = d.repeat,
                    writeSpeed = d.writeSpeed,
                    eraseSpeed = d.eraseSpeed,
                    pauseBetween = d.pauseBetween,
                    caretColor = d.caretColor,
                    caretWidth = d.caretWidth,
                    caretSpacing = d.caretSpacing,
                    blinkSpeed = d.blinkSpeed
                } = wordset as Wordset;

                const totalTime = animationLength(words, writeSpeed, eraseSpeed, pauseBetween);

                const animations = `, write-${esc(key, {
                    isIdentifier: true
                })} ${totalTime}s ${delay}s ${repeat !== -1 ? repeat + 1 : 'infinite'} forwards`;

                return {
                    [`.${esc(`type-${key}`, {
                        isIdentifier: true
                    })}`]: {
                        '&::after': {
                            content: '""',
                            'border-right': `${caretWidth} solid ${caretColor}`,
                            'padding-right': `${caretSpacing}`,
                            animation: `blink-caret-${key} ${blinkSpeed}s infinite${animations}`,
                            transition: 'border-color 50ms cubic-bezier(0.4, 0, 0.2, 1)'
                        }
                    }
                };
            })
        ];

        addUtilities(utilities);
    };
});

/*
 * Core Functions
 */

// Generates keyframes for each word
function generateKeyframes(
    label: string,
    words: string[],
    writeSpeed: number,
    eraseSpeed: number,
    pauseBetween: number,
    infinite: boolean
) {
    const keyframes = {
        from: { content: '""' },
        to: { content: `"${words[words.length - 1]}"` }
    };
    let lastKeyframe = 0;

    const totalTime = animationLength(words, writeSpeed, eraseSpeed, pauseBetween);

    words.forEach((currentWord: string) => {
        // Write animation
        const writeSteps = calculateSteps(currentWord, writeSpeed, totalTime);

        writeSteps.forEach((currentStep: number, i: number) => {
            if (currentStep) {
                currentStep = randomize(lastKeyframe + currentStep);

                if (i === writeSteps.length - 1) {
                    // Briefly pause animation on last keyframe
                    const startPause = currentStep;
                    const endPause = (lastKeyframe = startPause + (pauseBetween / totalTime) * 100);

                    // @ts-ignore
                    keyframes[`${round(startPause)}%, ${round(endPause)}%`] = {
                        content: `"${esc(currentWord.substring(0, i), { quotes: 'double' })}"`
                    };
                } else {
                    // @ts-ignore
                    keyframes[`${round(currentStep)}%`] = {
                        content: `"${esc(currentWord.substring(0, i), { quotes: 'double' })}"`
                    };
                }
            }
        });

        // Erase animation
        if (eraseSpeed) {
            const eraseSteps = calculateSteps(currentWord, eraseSpeed, totalTime);

            eraseSteps.forEach((currentStep: number, i: number) => {
                if (currentStep) {
                    currentStep = lastKeyframe + currentStep;
                    lastKeyframe = i === writeSteps.length - 1 ? currentStep : lastKeyframe;

                    // @ts-ignore
                    keyframes[`${round(currentStep)}%`] = {
                        content: `"${esc(currentWord.substring(0, currentWord.length - i), {
                            quotes: 'double'
                        })}"`
                    };
                }
            });
        }
    });

    if (infinite || eraseSpeed !== 0) {
        keyframes.to = { content: '""' };
    }

    return {
        [`@keyframes write-${esc(label, {
            isIdentifier: true
        })}`]: keyframes
    };
}

// Calculates the total animation time in seconds
function animationLength(
    words: string[],
    writeSpeed: number,
    eraseSpeed: number,
    pauseBetween: number
) {
    let totalTime = 0;
    words.forEach((currentWord: string) => {
        totalTime +=
            (Number(writeSpeed) + Number(eraseSpeed)) * currentWord.length + Number(pauseBetween);
    });

    return totalTime;
}

// Calculates keyframe steps for word animation phase (e.g. write, erase)
function calculateSteps(currentWord: string, phaseSpeed: number, totalTime: number) {
    const phaseTime = Number(phaseSpeed) * currentWord.length;
    const phasePercentage = (phaseTime / Number(totalTime)) * 100;
    return num.quantize(num.interpolate(0, phasePercentage), currentWord.length + 1);
}

/*
 * Helper Functions
 */

// Shorthand for rounding a value 2 decimal places
function round(number: number) {
    return Number(Number(number).toFixed(2));
}

// Add a small random variance to a number
function randomize(number: number, variance: number = 0.5) {
    const max = Number(number) + Number(variance);
    const min = Number(number) - Number(variance);

    return Math.random() * (max - min) + min;
}
